pkg/farm/list_builder.go

```diff
@@ -0,0 +1,320 @@
+			pullOptions := entities.PullToFileOptions{
+				ImageID:    image.ID,
+				SaveFormat: image.SaveFormat,
+				SaveFile:   tempFile.Name(),
+			}
+			if image.SaveFormat == manifest.DockerV2Schema2MediaType {
+				listFormat = manifest.DockerV2ListMediaType
+				imageFormat = manifest.DockerV2Schema2MediaType
+			}
+			reference, err := engine.PullToFile(ctx, pullOptions)
+			if err != nil {
+				return fmt.Errorf("pulling image %q to temporary directory: %w", image, err)
+			}
+			ref, err := alltransports.ParseImageName(reference)
+			if err != nil {
+				return fmt.Errorf("pulling image %q to temporary directory: %w", image, err)
+			}
+			refsMutex.Lock()
+			defer refsMutex.Unlock()
+			refs[image] = ref
+			return nil
+		})
+	}
+	pullErrors = pullGroup.Wait()
+	err = pullErrors.ErrorOrNil()
+	if err != nil {
+		return "", fmt.Errorf("building: %w", err)
+	}
+
+	if m.options.cleanup {
+		var rmGroup multierror.Group
+		for image, engine := range images {
+			image, engine := image, engine
+			rmGroup.Go(func() error {
+				_, err := engine.Remove(ctx, []string{image.ID}, entities.ImageRemoveOptions{})
+				if len(err) > 0 {
+					return err[0]
+				}
+				return nil
+			})
+		}
+		rmErrors := rmGroup.Wait()
+		if rmErrors != nil {
+			if err = rmErrors.ErrorOrNil(); err != nil {
+				return "", fmt.Errorf("removing intermediate images: %w", err)
+			}
+		}
+	}
+
+	supplemental := []types.ImageReference{}
+	var (
+		sys     types.SystemContext
+		mDigest digest.Digest
+	)
+	// Create a manifest list
+	list := lmanifests.Create()
+	// Add the images to the list
+	for image, ref := range refs {
+		if mDigest, err = list.Add(ctx, &sys, ref, true); err != nil {
+			return "", fmt.Errorf("adding image %q to list: %w", image.ID, err)
+		}
+		supplemental = append(supplemental, ref)
+	}
+	// Save the list to the temporary directory to be the main manifest
+	listBytes, err := list.Serialize(listFormat)
+	if err != nil {
+		return "", fmt.Errorf("serializing manifest list: %w", err)
+	}
+	if err = os.WriteFile(filepath.Join(tempDir, "manifest.json"), listBytes, fs.FileMode(0o600)); err != nil {
+		return "", fmt.Errorf("writing temporary manifest list: %w", err)
+	}
+
+	// Now copy everything to the final dir: location
+	defaultPolicy, err := signature.DefaultPolicy(&sys)
+	if err != nil {
+		return "", err
+	}
+	policyContext, err := signature.NewPolicyContext(defaultPolicy)
+	if err != nil {
+		return "", err
+	}
+	input := supplemented.Reference(tempRef, supplemental, cp.CopyAllImages, nil)
+	copyOptions := cp.Options{
+		ForceManifestMIMEType: imageFormat,
+		ImageListSelection:    cp.CopyAllImages,
+	}
+	_, err = cp.Image(ctx, policyContext, output, input, &copyOptions)
+	if err != nil {
+		return "", fmt.Errorf("copying images to dir:%q: %w", m.directory, err)
+	}
+
+	// Write the manifest list's ID file if we're expected to
+	if m.options.iidFile != "" {
+		if err := os.WriteFile(m.options.iidFile, []byte(mDigest), 0644); err != nil {
```

Comment:
`mDigest` is the digest of the manifest list, not an image ID with "sha256:" hard-prefixed to it for compatibility reasons.  I think it would be better if this version didn't accept `--iidfile`.